src/axom/primal/geometry/KnotVector.hpp

```diff
@@ -0,0 +1,501 @@
+    // Load basis functions
+    for(int j = 0; j <= m_deg; j++)
+    {
+      ders[0][j] = ndu[j][m_deg];
+    }
+
+    // This section computes the derivatives (Eq. [2.9])
+
+    // Loop over function index.
+    for(int r = 0; r <= m_deg; r++)
+    {
+      int s1 = 0, s2 = 1;  // Alternate rows in array a
+      a[0][0] = 1.;
+      // Loop to compute kth derivative
+      for(int k = 1; k <= n; k++)
+      {
+        T d = 0.;
+        int rk = r - k;
+        int pk = m_deg - k;
+        if(r >= k)
+        {
+          a[s2][0] = a[s1][0] / ndu[pk + 1][rk];
+          d = a[s2][0] * ndu[rk][pk];
+        }
+        int j1 = (rk >= -1) ? 1 : -rk;
+        int j2 = (r - 1 <= pk) ? (k - 1) : (m_deg - r);
+        for(int j = j1; j <= j2; j++)
+        {
+          a[s2][j] = (a[s1][j] - a[s1][j - 1]) / ndu[pk + 1][rk + j];
+          d += a[s2][j] * ndu[rk + j][pk];
+        }
+        if(r <= pk)
+        {
+          a[s2][k] = -a[s1][k - 1] / ndu[pk + 1][r];
+          d += a[s2][k] * ndu[r][pk];
+        }
+        ders[k][r] = d;
+        // Switch rows
+        std::swap(s1, s2);
+      }
+    }
+    // Multiply through by the correct factors (Eq. [2.9])
+    T r = static_cast<T>(m_deg);
+    for(int k = 1; k <= n; k++)
+    {
+      for(int j = 0; j <= m_deg; j++)
+      {
+        ders[k][j] *= r;
+      }
+      r *= static_cast<T>(m_deg - k);
+    }
+  }
+
+  /// \brief Return the degree of the knot vector
+  int getDegree() const { return m_deg; }
+
+  /// \brief Return the number of knots in the knot vector
+  axom::IndexType getNumKnots() const { return m_knots.size(); }
+
+  /// \brief Return the number of control points implied by the knot vector
+  axom::IndexType getNumControlPoints() const
+  {
+    return m_knots.size() - m_deg - 1;
+  }
+
+  /// \brief Clear the list of knots
+  void clear()
+  {
+    m_deg = -1;
+    m_knots.clear();
+  }
+
+  /*!
+   * \brief Returns the index of the knot span containing parameter t
+   * 
+   * Implementation adapted from Algorithm A2.1 on page 68 of "The NURBS Book"
+   */
+  axom::IndexType findSpan(double t) const
+  {
+    const axom::IndexType nkts = m_knots.size();
+
+    if(t <= m_knots[0])
+    {
+      return m_deg;
+    }
+
+    if(t >= m_knots[nkts - 1])
+    {
+      return nkts - m_deg - 2;
+    }
+
+    // perform binary search on the knots,
+    axom::IndexType low = m_deg;
+    axom::IndexType high = nkts - m_deg - 1;
+    axom::IndexType mid = (low + high) / 2;
+    while(t < m_knots[mid] || t >= m_knots[mid + 1])
+    {
+      (t < m_knots[mid]) ? high = mid : low = mid;
+      mid = (low + high) / 2;
+    }
+
+    return mid;
+  }
+
+  /// \brief Returns the index of the knot span and the multiplicity
+  axom::IndexType findSpan(double t, int& multiplicity)
+  {
+    const auto nkts = m_knots.size();
+    const auto span = findSpan(t);
+
+    multiplicity = 0;
+    for(auto i = (t == m_knots[nkts - 1]) ? nkts - 1 : span; i >= 0; --i)
+    {
+      if(m_knots[i] == t)
```

Comment:
Yes, I think it should check strict equality here. Otherwise you open a whole can of worms about what constitutes a clamped curve, i.e. is a 2nd degree curve with a knot vector `{-1e-8, 0, 1e-8, 0.5, 1, 1, 1}` a valid curve?